# G76 Program Scraping

- This notebook is to document the scraping , cleaning, formatting, and merging processes done to get the G76 code information with the correct drawing number and material type.

In [607]:
# import dependencies
import os

# create an empty dictionary
G76_dict = {}
# define root directory to search
root = r'\\REST-SAVEIT\\Users\\TaylorR\\OLD SAVEIT2\\CNC Master File\\ALL SWISS MACHINES\\'
# use 'os.walk' to check sub directories
for path, subdirs, files in os.walk(root + '\\'):
        # iterate thru checking the file names
        for name in files:
                # join the file name with it's raw path into variable 'directory'
                directory = os.path.join(path, name)
                # every program will have two G76 lines, this will start to address that
                new_name = (name + '_block_1')
                # all 'if not statements' below are checking file types just to save time in iteration
                if not directory.lower().endswith('.pdf'):
                        if not directory.lower().endswith('.doc'):
                                if not directory.lower().endswith('.part'):
                                        if not directory.lower().endswith('.ppt'):
                                                if not directory.lower().endswith('.dbp'):
                                                        if not directory.lower().endswith('.back'):
                                                                if not directory.lower().endswith('.cpd'):
                                                                        if not directory.lower().endswith('.trf'):
                                                                                if not directory.lower().endswith('.db'):
                                                                                        # open the directory's file in read only mode
                                                                                        o = open(directory, 'r', encoding='Latin-1')
                                                                                        # splitting the directory variable to catpure machine type
                                                                                        s = os.path.split(directory)
                                                                                        a = s[1]
                                                                                        b = s[0]
                                                                                        # same for drawing number
                                                                                        s = os.path.split(b)
                                                                                        c = s[1] 
                                                                                        b = s[0]
                                                                                        s = os.path.split(b)
                                                                                        # same for program number
                                                                                        e = s[1]
                                                                                        b = s[0]
                                                                                        # read through the lines of text one by one
                                                                                        for line in o.readlines():
                                                                                                # make the line lower case for processessing
                                                                                                line = str(line.lower())
                                                                                                line = line.strip().lower()
                                                                                                # look for string
                                                                                                if 'g76' in line:
                                                                                                        # if first G76 line already added to the dictionary
                                                                                                        # then change to 'block 2'
                                                                                                        if new_name in G76_dict:
                                                                                                                G76_dict.update({name + '_block_2' : [line, c, e]})
                                                                                                        else:
                                                                                                                # else add 'block 1'
                                                                                                                G76_dict.update({new_name : [line, c, e]})

# same as function above but with a different root
root = r'\\REST-SAVEIT\\Users\\TaylorR\\OLD SAVEIT2\\CNC Master File\ALL HARDINGE\\'
for path, subdirs, files in os.walk(root + '\\'):
        for name in files:
                directory = os.path.join(path, name)
                new_name = (name + '_block_1')
                if not directory.lower().endswith('.pdf'):
                        if not directory.lower().endswith('.doc'):
                                if not directory.lower().endswith('.part'):
                                        if not directory.lower().endswith('.ppt'):
                                                if not directory.lower().endswith('.dbp'):
                                                        if not directory.lower().endswith('.back'):
                                                                if not directory.lower().endswith('.cpd'):
                                                                        if not directory.lower().endswith('.trf'):
                                                                                if not directory.lower().endswith('.db'):
                                                                                        o = open(directory, 'r', encoding='Latin-1')
                                                                                        s = os.path.split(directory)
                                                                                        a = s[1]
                                                                                        b = s[0]
                                                                                        s = os.path.split(b)
                                                                                        c = s[1] 
                                                                                        b = s[0]
                                                                                        s = os.path.split(b)
                                                                                        e = s[1]
                                                                                        b = s[0]
                                                                                        for line in o.readlines():
                                                                                                line = str(line.lower())
                                                                                                line = line.strip().lower()
                                                                                                if 'g76' in line:
                                                                                                        if new_name in G76_dict:
                                                                                                                G76_dict.update({name + '_block_2' : [line, c, e]})
                                                                                                        else:
                                                                                                                G76_dict.update({new_name : [line, c, e]})
# write the dictionary to a .csv file
myfile = 'file.csv'
with open(myfile, 'w') as f:
        for key, value in G76_dict.items():
                f.write('%s, %s\n' % (key, value))

### Import Data

In [655]:
# import dependencies
import pandas as pd

# window formatting for viewing
pd.set_option('display.max_columns', None)  
pd.set_option('display.max_rows', None)

# create a data frame from the scraped CNC programs
df = pd.read_csv('C:\\Users\\MacalusoC\\Desktop\\Technical Docs\\Thread Data Scrape\\file.csv', header = None)
# create a header w/ column names
df.columns = ['Program Number', 'G76 Code', 'Drawing Number', 'Machine']

# import the .xlsx file with part number and material taken from MS Access 'SetupSheets' db
dfMaterial = pd.read_excel('C:\\Users\\MacalusoC\\Desktop\\Technical Docs\\Thread Data Scrape\\PartData.xlsx')
# rename the columns
dfMaterial.columns = ['Drawing Number', 'Material']

In [656]:
# look at some scraped data
df.head()

Program Number                           G76 Code  \
0  O0001-HEAD#1.nc_block_1               ['g76p040060q0015r0'   
1  O0001-HEAD#1.nc_block_2   ['g76x.325z.525r0p230q230f.0416'   
2  O0004-HEAD#1.nc_block_1               ['g76p040060q0015r0'   
3  O0004-HEAD#1.nc_block_2     ['g76x.377z1.12r0p310q310f.05'   
4  O0009-HEAD#1.nc_block_1               ['g76p040060q0015r0'   

                   Drawing Number     Machine  
0   '0001-1001 JEFD0389130.REV-A'   'KNC-32']  
1   '0001-1001 JEFD0389130.REV-A'   'KNC-32']  
2   '0004-1004 BASD0500310.REV-A'   'KNC-32']  
3   '0004-1004 BASD0500310.REV-A'   'KNC-32']  
4   '0009-1009 JEFD0304260.REV-B'   'KNC-32']

In [657]:
# look at some imported table data
dfMaterial.head()

Drawing Number                Material
0  AARD0500210 C                     303
1  AARD0504150 A                  15-5PH
2  AARD0504160 A                  15-5PH
3  AARD0504960 B  SB-425 N08825 ANNEALED
4  AARD0505750 A  SB-425 N08825 ANNEALED

In [658]:
# look at the shape (rows vs columns) of data
print('The MS Access table has ' + str(dfMaterial.shape[0]) + ' rows and ' + str(dfMaterial.shape[1]) + ' columns.')
print('The scraped G76 data has ' + str(df.shape[0]) + ' rows and ' + str(df.shape[1]) + ' columns.')
print('Continueing below this will be displayed as the code "dataframe_name".shape, which outputs "row, column" sizes')

The MS Access table has 651 rows and 2 columns.
The scraped G76 data has 1552 rows and 4 columns.
Continueing below this will be displayed as the code "dataframe_name".shape, which outputs "row, column" sizes


### Format 'Machine' Column
- There's a ton of columns that do not equal a real machine. This is from the windows explorer directory at 'root' above be structured differently from part to part. 
- For this exerise it will not be corrected. There will be fall out of data throughout the cleaning process and this is the first.

In [659]:
#List unique values in the'Machine' column
df.Machine.unique()

array([" 'KNC-32']", " 'SA-16']", " 'ALL SWISS MACHINES']", " 'SV-20']",
       " 'SDR1326A-4 REV-A']", " '301 BACKUP']", " 'T-BO-326 TSUGAMI']",
       " '5016-5017 AARD0511110M REV B']",
       " '5018-5019 AARD0511030M REV B']", " '302 BACKUP']",
       " 'T-S-206 TSUGAMI']", " 'TSUGAMI PATH 1']", " 'TSUGAMI PATH 2']",
       " 'CHNC III']", " 'JEFD0322970 REV A']", " 'VXLD0301010 REV J']",
       " 'T-42']", " 'GT']", " 'JEFD0301230 REV A']", " 'RENUMBERING']",
       " 'VDLD0302060 REV. C']", " 'GT ROBOT']", " 'NTY_3 Files']",
       " 'AARD0517260 REV B']", " 'JEFD0302000 REV E']",
       " 'JEFD0303400 REV G']", " 'JEFD0304381 REV A']",
       " 'ALL HARDINGE']", " 'REUSE']", " 'TWIN TURN']",
       " 'JEFD0303900 REV A']", " 'VDLD0501070 REV B']",
       " 'JEFD0365530 REV A']", " 'JELD0508460 REV A']",
       " 'JELD0508510 REV B']", " 'JELD0508560 REV A']",
       " 'JELD0508610 REV B']", " 'WY-100']", " 'BUILDING BLOCKS']",
       " 'CONTROL CLEAN-OUT 05-18']", " 'JEFD031732

In [660]:
# create a list of real machine folder names from above
machineList = [" 'CHNC III']",
               " 'GT ROBOT']",
               " 'GT']",
               " 'KNC-32']",
               " 'NTY_3 Files']",
               " 'SA-16']",
               " 'SV-20']",
               " 'T-42']",
               " 'T-BO-326 TSUGAMI']",
               " 'T-S-206 TSUGAMI']",
               " 'TWIN TURN']",
               " 'WY-100']",
               " 'WY-150]'", ]
# remove any rows that don't contain a value from the list above in 'Machine'
df = df[df['Machine'].isin(machineList)]
# rename the 'Machine' string values to what they're supposed to be
df['Machine'] = df['Machine'].replace({" 'CHNC III']":'CHNC III', 
                                       " 'GT ROBOT']": 'GT ROBOT', 
                                       " 'GT']": 'GT',
                                       " 'KNC-32']": 'KNC-32',
                                       " 'NTY_3 Files']": 'NTY-3',
                                       " 'SA-16']": 'SA-16',
                                       " 'SV-20']": 'SV-20',
                                       " 'T-42']": 'T-42',
                                       " 'T-BO-326 TSUGAMI']": 'BO-326',
                                       " 'T-S-206 TSUGAMI']": ' S-206',
                                       " 'TWIN TURN']": 'TWIN TURN',
                                       " 'WY-100']": 'WY-100',
                                       " 'WY-150]'": 'WY-150'})
# check the output
df.Machine.unique()

array(['KNC-32', 'SA-16', 'SV-20', 'BO-326', ' S-206', 'CHNC III', 'T-42',
       'GT', 'GT ROBOT', 'NTY-3', 'TWIN TURN', 'WY-100'], dtype=object)

In [661]:
# looks like we lost about 400 rows of data, or 200 G76 blocks so far
df.shape

(1194, 4)

In [662]:
# assign 'Block' column value based on pattern match from 'Program Number'
block1 = df['Program Number'].str.contains('block_1')
# replace the boolian return of the above line with '1' or '2' for block
df['Block'] = block1.replace((True,False), ('1','2'))
df.head(6)

Program Number                           G76 Code  \
0  O0001-HEAD#1.nc_block_1               ['g76p040060q0015r0'   
1  O0001-HEAD#1.nc_block_2   ['g76x.325z.525r0p230q230f.0416'   
2  O0004-HEAD#1.nc_block_1               ['g76p040060q0015r0'   
3  O0004-HEAD#1.nc_block_2     ['g76x.377z1.12r0p310q310f.05'   
4  O0009-HEAD#1.nc_block_1               ['g76p040060q0015r0'   
5  O0009-HEAD#1.nc_block_2   ['g76x.495z.465r0p320q320f.0555'   

                   Drawing Number Machine Block  
0   '0001-1001 JEFD0389130.REV-A'  KNC-32     1  
1   '0001-1001 JEFD0389130.REV-A'  KNC-32     2  
2   '0004-1004 BASD0500310.REV-A'  KNC-32     1  
3   '0004-1004 BASD0500310.REV-A'  KNC-32     2  
4   '0009-1009 JEFD0304260.REV-B'  KNC-32     1  
5   '0009-1009 JEFD0304260.REV-B'  KNC-32     2

In [663]:
# clean the 'Drawing Number' columnn
char_list = ["'",
             ' REV',
             '.',]
# loop through the list above
for char in char_list:
    # replace the character with nothing, esseentially deleting it
    df['Drawing Number'] = df['Drawing Number'].str.replace(char, '')
# replace other strings
df['Drawing Number'] = df['Drawing Number'].str.replace('REV-', ' ')
df['Program Number'] = df['Program Number'].str.replace('_block_1', '')
df['Program Number'] = df['Program Number'].str.replace('_block_2', '')
# clean 'G76 Code' column
char_list = ['[',
             "'",
             '/',]
for char in char_list:
    df['G76 Code'] = df['G76 Code'].str.replace(char, '')
# uppcase characters in 'G76 Code' column
df['G76 Code'] = df['G76 Code'].str.upper()

In [664]:
# iterate the df
for index, row in df.iterrows():
    # find a '-' in 'Drawing Number' as 5th character, for strings like "xxxx-xxxx DTCD0100070 A"
    if row[2].find('-') == 5:
        # remove the unwanted leading characters
        row[2] = row[2][10:]
    # remove the whitespace before and after string in 'Program Name':
    row[2] = row[2].rstrip(' ')
    row[2] = row[2].lstrip(' ')
df.head()
#df.shape

Program Number                        G76 Code Drawing Number Machine  \
0  O0001-HEAD#1.nc               G76P040060Q0015R0  JEFD0389130 A  KNC-32   
1  O0001-HEAD#1.nc   G76X.325Z.525R0P230Q230F.0416  JEFD0389130 A  KNC-32   
2  O0004-HEAD#1.nc               G76P040060Q0015R0  BASD0500310 A  KNC-32   
3  O0004-HEAD#1.nc     G76X.377Z1.12R0P310Q310F.05  BASD0500310 A  KNC-32   
4  O0009-HEAD#1.nc               G76P040060Q0015R0  JEFD0304260 B  KNC-32   

  Block  
0     1  
1     2  
2     1  
3     2  
4     1

In [665]:
df.shape

(1194, 5)

In [666]:
#df = df.groupby(['Program Number', 'Drawing Number', 'Machine'])['G76 Code'].apply(', '.join)
#df.shape

In [667]:
# grouping every two rows based on Program Number, Drawing Number, and Machine to get the two G76 lines on the same row of data
df = df.groupby('Program Number').agg({'Drawing Number':'first', 
                             'G76 Code': ', '.join, 
                             'Machine':'first' }).reset_index()
df.head()

Program Number                Drawing Number  \
0       0002-TTS  X - FILES NOT RUN SINCE 2012   
1       0003-TTM  X - FILES NOT RUN SINCE 2012   
2        0004-GT  X - FILES NOT RUN SINCE 2012   
3       0004-TTS  X - FILES NOT RUN SINCE 2012   
4       0005-TTM  X - FILES NOT RUN SINCE 2012   

                                            G76 Code    Machine  
0   G76P020060Q0020R.0005,  G76X.7566Z-.6000P0338...  TWIN TURN  
1   G76P030060Q0025R.0005,  G76X.3767Z-.38P0306Q0...  TWIN TURN  
2   G76P020060Q00200R.0005,  G76X.3767Z-.48P03067...         GT  
3   G76P030060Q0025R.0005,  G76X.3767Z.38P0306Q03...  TWIN TURN  
4   G76P020060Q0025R.0005,  G76X.3767Z-.38P0306Q0...  TWIN TURN

In [668]:
df.shape

(617, 4)

### Merge data frames

In [669]:
# merge the two dataframes, scraped data & MS Access table data
# they'e being merged on the 'Drawing Number' column, this is another step that will
# have a lot of fall out due to the directory inconsistency from earlier
df_new = pd.merge(df, dfMaterial, how = 'left', on = 'Drawing Number')
# show a sample
df_new.sample(6)

Program Number                    Drawing Number  \
547        O5017.P-2  OFF LOADED FROM MACHINE 09-04-15   
512  O1072-HEAD#2.nc                     JEFD0317570 A   
52         00628-TTS                      card 2-27-14   
82          0144-TTS      X - FILES NOT RUN SINCE 2012   
290         0770-TTS                     FLLD0500550 B   
549        O5019.P-2  OFF LOADED FROM MACHINE 09-04-15   

                                              G76 Code    Machine Material  
547   G18G76P010060Q00250R.0005,  G76X.3173Z-.477P0...     BO-326      NaN  
512       G76P021030Q10R0,  G76X.438Z.44R0P290Q200F.05     KNC-32      NaN  
52    G76P020060Q0025R.0005,  G76X.3767Z.49P0306Q03...       T-42      NaN  
82    G76P050060Q0010R.0005,  G76X.375Z.19P0225Q022...  TWIN TURN      NaN  
290   G76P020060Q0025R.0005,  G76X.5055Z.355P0271Q0...  TWIN TURN      303  
549   G18G76P010060Q00250R.0005,  G76X.3691Z-.37P03...     BO-326      NaN

In [670]:
df_new.shape

(617, 5)

### Final cleaning and formatting

In [671]:
# drop any rows that contain a 'NaN' value (not a number)
df_new.dropna(inplace=True)
df_new.shape

(251, 5)

In [672]:
df_new.head(251)

Program Number  Drawing Number  \
6                    0007-TTM   JEFD0304260 B   
7                    0008-TTS   JEFD0304260 B   
11                   0013-TTM   JEFD0304002 D   
12                   0014-TTS   JEFD0304002 D   
16                   0019-TTM   JEFD0304002 D   
17                   0020-TTS   JEFD0365530 A   
22        0025-TTM HEX UPDATE   JEBD0507160 A   
24   0025-TTM ORG ROUND STOCK   JEBD0507160 A   
25        0026-TT5 HEX UPDATE   JEBD0507160 A   
27   0026-TTS ORG ROUND STOCK   JEBD0507160 A   
39                  0041-SA16   JETD0300380 B   
43                   0045-TTM   JEFD0303400 G   
44                   0046-TTS   JEFD0303400 G   
45                   0048-TTS   VDLD0501090 B   
46                   0049-T42   JEFD0317400 A   
53                   0067-T42   JEFD0319390 A   
61                   0110-TTS   JELD0508610 B   
62                   0114-TTS   JELD0508460 A   
65                   0117-TTM   VXLD0301030 H   
66                   0118-TTS   VXLD0301030 H   
76                   0130-TTS   VDLD0302261 B   
78                   0137-TTM   JEFD0317150 A   
79                   0138-TTS   JEFD0317150 A   
96                   0175-TTM   JEFD0317060 A   
98                   0176-TTS   JEFD0317060 A   
102                  0181-TTM   JEFD0302410 A   
103                  0182-TTS   JEFD0302410 A   
104                  0183-TTM   JEFD0302390 A   
105                  0184-TTS   JEFD0302390 A   
112                  0193-TTM   JEFD0302470 C   
113                  0194-TTS   JEFD0302470 C   
123                  0209-3SP   JEBD0507160 A   
124                  0209-5SP   JEBD0507160 A   
125                  0210-3SP   JEBD0507160 A   
131                  0218-3SP   JEFD0302000 E   
132                  0219-3SP   JEFD0302000 E   
133             0219-3SP REAM   JEFD0302000 E   
134                  0221-3SP   JEFD0302330 B   
135                  0222-3SP   JEFD0302330 B   
138                  0225-3SP   JEFD0303400 G   
139                  0226-3SP   JEFD0303400 G   
140                  0227-3SP   JEFD0303420 B   
141                  0228-3SP   JEFD0303420 B   
148                  0236-3SP   JEFD0304002 D   
149                  0237-3SP   JEFD0304002 D   
150                  0238-3SP   JEFD0304011 D   
151                  0239-3SP   JEFD0304011 D   
152                  0240-3SP   JEFD0304040 B   
153                  0241-3SP   JEFD0304040 B   
154                  0245-2SP   JEFD0304381 A   
155                  0245-3SP   JEFD0304381 A   
156                  0246-3SP   JEFD0304381 A   
171                  0272-3SP   JEFD0365500 A   
172                  0273-3SP   JEFD0365500 A   
180                  0287-3SP   REFD0501310 B   
181                  0288-3SP   REFD0501310 B   
188                  0318-3SP   VDCD0305010 G   
193                  0341-3SP   VXLD0301000 F   
224                  0412-3SP   JEFD0302410 A   
225                  0413-3SP   JEFD0302410 A   
235                  0508-3SP   JEFD0302570 A   
236                  0509-3SP   JEFD0302570 A   
237                  0510.P-1   VXLD0500510 C   
238                  0510.P-2   VXLD0500510 C   
239               0510_P1.txt   VXLD0500510 C   
240               0510_P2.txt   VXLD0500510 C   
244                   0566-GT   JHBD0200010 C   
245                   0576-GT   JEFD0303400 G   
246                   0580-GT   JEFD0301230 A   
263                  0693-TTM   JEFD0365500 A   
264                  0694-TTS   JEFD0365500 A   
267                  0708-3SP   VDLD0501090 B   
270                  0720-TTM   JEFD0302430 A   
271                  0721-TTS   JEFD0302430 A   
274                  0738-3SP   JEFD0317600 A   
275                  0739-3SP   JEFD0317600 A   
278                  0748-T42   VDLD0302250 B   
279                  0749-TTM   JEFD0302590 A   
282                  0761-TTM   JEFD0302450 B   
283                  0762-TTS   JEFD0302450 B   
284                  0763-3SP   JEFD0302450 B 

In [673]:
# iterate over the dataframe
for index, row in df_new.iterrows():
    # strip white space on left of row value
    row[2] = row[2].lstrip(' ')
    # if the G76 column doesn't start with G
    if (str(row[2])[0]) != 'G':
        # drop if it
        df_new.drop(index, inplace = True)
# because of initial rows being dropped, certain observations could not be found. Required it's own iterrow()
for index, row in df_new.iterrows():        
    # check for ',' in G76 Code column
    if ',' not in str(row[2]):
        # drop if it doesn't exist
        df_new.drop(index, inplace = True)
df_new.shape

(233, 5)

In [674]:
# split 'G76 Code' column by ','
split = df_new['G76 Code'].str.split(',', n = 1, expand = True) 
# set first split to new column 
df_new['G76 Block 1']= split[0] 
# set second split to second new column
df_new['G76 Block 2']= split[1] 
# drop the original column
df_new.drop(columns =['G76 Code'], inplace = True) 
# reorganize columns
df_new = df_new[[ 'Drawing Number', 'Material', 'Program Number', 'G76 Block 1', 'G76 Block 2', 'Machine']]
df_new.head(233)

Drawing Number         Material            Program Number  \
6     JEFD0304260 B             304L                  0007-TTM   
7     JEFD0304260 B             304L                  0008-TTS   
11    JEFD0304002 D              304                  0013-TTM   
12    JEFD0304002 D              304                  0014-TTS   
16    JEFD0304002 D              304                  0019-TTM   
17    JEFD0365530 A              304                  0020-TTS   
22    JEBD0507160 A             304L       0025-TTM HEX UPDATE   
24    JEBD0507160 A             304L  0025-TTM ORG ROUND STOCK   
25    JEBD0507160 A             304L       0026-TT5 HEX UPDATE   
27    JEBD0507160 A             304L  0026-TTS ORG ROUND STOCK   
39    JETD0300380 B              304                 0041-SA16   
43    JEFD0303400 G              304                  0045-TTM   
44    JEFD0303400 G              304                  0046-TTS   
45    VDLD0501090 B             304L                  0048-TTS   
46    JEFD0317400 A              304                  0049-T42   
53    JEFD0319390 A           6AL-4V                  0067-T42   
62    JELD0508460 A             304L                  0114-TTS   
65    VXLD0301030 H             304L                  0117-TTM   
66    VXLD0301030 H             304L                  0118-TTS   
76    VDLD0302261 B             304L                  0130-TTS   
78    JEFD0317150 A              304                  0137-TTM   
79    JEFD0317150 A              304                  0138-TTS   
96    JEFD0317060 A              304                  0175-TTM   
98    JEFD0317060 A              304                  0176-TTS   
102   JEFD0302410 A              304                  0181-TTM   
103   JEFD0302410 A              304                  0182-TTS   
104   JEFD0302390 A              304                  0183-TTM   
105   JEFD0302390 A              304                  0184-TTS   
112   JEFD0302470 C           15-5PH                  0193-TTM   
113   JEFD0302470 C           15-5PH                  0194-TTS   
123   JEBD0507160 A             304L                  0209-3SP   
124   JEBD0507160 A             304L                  0209-5SP   
125   JEBD0507160 A             304L                  0210-3SP   
131   JEFD0302000 E             304L                  0218-3SP   
132   JEFD0302000 E             304L                  0219-3SP   
133   JEFD0302000 E             304L             0219-3SP REAM   
134   JEFD0302330 B             304L                  0221-3SP   
135   JEFD0302330 B             304L                  0222-3SP   
138   JEFD0303400 G              304                  0225-3SP   
139   JEFD0303400 G              304                  0226-3SP   
140   JEFD0303420 B              304                  0227-3SP   
141   JEFD0303420 B              304                  0228-3SP   
148   JEFD0304002 D              304                  0236-3SP   
149   JEFD0304002 D              304                  0237-3SP   
150   JEFD0304011 D              304                  0238-3SP   
151   JEFD0304011 D              304                  0239-3SP   
152   JEFD0304040 B             304L                  0240-3SP   
153   JEFD0304040 B             304L                  0241-3SP   
154   JEFD0304381 A             304L                  0245-2SP   
155   JEFD0304381 A             304L                  0245-3SP   
156   JEFD0304381 A             304L                  0246-3SP   
171   JEFD0365500 A              304                  0272-3SP   
172   JEFD0365500 A              304                  0273-3SP   
180   REFD0501310 B           6AL-4V                  0287-3SP   
181   REFD0501310 B           6AL-4V                  0288-3SP   
188   VDCD0305010 G              303                  0318-3SP   
193   VXLD0301000 F              304                  0341-3SP   
224   JEFD0302410 A              304                  0412-3SP   
225   JEFD0302410 A              304                  0413-3SP   
235   JEFD0302570 A           15-5PH         

### Output data to CSV

In [675]:
df_new.to_csv('G76_Data.csv')

### Ending notes
- the initial scraping has 1552 rows captured containing a string of 'G76', that's a potential 776 programs (two blocks each) that could have been referenced. 
- Due to inconsistent naming or structuring or directories a majority of these were dropped throughout the cleaning process (543 program worth of G76 data).
- A more rigorous function could be written to capture the needed information in order to not lose so much data but for this initial exercise it didn't seem worth it.